### ISSUE TITLE
C/C++ Go to/Peek definition not working and parsing is stuck

### ISSUE THREAD

**Issue by weinand** [microsoft/vscode-cpptools#962]
_From @hagaishapira on August 9, 2017 14:36_

- VSCode Version: Code 1.14.2 (cb82febafda0c8c199b9201ad274e25d9a76874e, 2017-07-19T23:26:08.116Z)
- OS Version: Linux x64 4.8.0-53-generic
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
meson|asa|1.0.0
python|don|0.7.0
cpptools|ms-|0.12.2
vim|vsc|0.9.0

(1 theme extensions excluded)

Opening my project on a new instance of code non of the navigation to definition / declaration work and all result in "No definition found for <Symbol>". 
Little read flame in the bottom right corner gets stuck after a while saying "Parsing..."

---

Reproduces without extensions:  No (extension related)

_Copied from original issue: Microsoft/vscode#32191_

**Comment by weinand** [comment]
_From @hagaishapira on August 9, 2017 14:50_

however, definition does show upon hovering.

**Comment by sean-mcmanus** [comment]
@hagaishapira Can you do a "C/Cpp: Edit Configuration" command and make sure your browse.path setting is correct? What is the Microsoft.VSCode.CPP.Extension.linux process doing? That process needs to locate and parse the files for the symbols to be found. If you manually open the file with the definition does Go to Definition start working? Does Go to Definition work with a simpler workspace? Does it only fail on certain variables? We might need help reproing the problem.

**Comment by horvay** [comment]
I wonder if this is also related to #928

**Comment by sean-mcmanus** [comment]
@horvay I think https://github.com/Microsoft/vscode-cpptools/issues/928 is different -- it's a perf issue with the Microsoft.VSCode.CPP.IntelliSense.Msvc process, but Go to Definition is not implemented by that process (it's implemented by the Tag Parser engine still).

**Comment by hagaishapira** [comment]
My browse.path is correct, but if my browse.databaseFilename is an empty string do you have a default path?
Opening a file manually doesn't seem to make it work.
I see I have multiple Microsoft.VSCode.CPP.Extension.linux processes running, some from a few days ago and they don't terminate upon vscode exit.

**Comment by sean-mcmanus** [comment]
@hagaishapira You should kill the other extension processes -- the symbol database won't function correctly if multiple processes are trying to access it (i.e. Go to Def would fail). Also, you can try adding a new databaseFilename, such as "${workspaceRoot}/.vscode/vc.db" to have the symbol engine regenerate. Then you can easily see how the size increases.

**Comment by hagaishapira** [comment]
@sean-mcmanus Great! that fixed it. Size of the db is stable and features available again. The Microsoft.VSCode.CPP.Extension.linux also terminates correctly. Any idea why the zombies happened in the first place?

**Comment by hagaishapira** [comment]
They all seemed to be waiting on a futex() call.

**Comment by sean-mcmanus** [comment]
@hagaishapira Unfortunately, we don't know why the processes doesn't exit sometimes. We're not able to repro the problem. However, we should add better error handling and telemetry for when the database can't be opened due to another processing using it.

**Comment by xueyunquan** [comment]
@sean-mcmanus : I personally recommend and suggest you guys try working on linux kernel with VS Code+cpptools. With such a large, complex project, the vscode is highly stressed and could potentially produce/reproduce more issues. Alternatively you may try gcc, llvm or ROCm. You may also find more ways to improve vscode.

**Comment by yyl123hf** [comment]
If the project folder contains the pathname of the Chinese character, then the go to define prompt can not find the definition.

**Comment by hagaishapira** [comment]
@sean-mcmanus Just an update, the problem keeps reoccurring. After some time the red parsing flame remains frozen. Microsoft.VSCode.CPP.Extension.linux remains running after vscode exits. Any way for me to provide you more info the next time this happens? Would attaching a debugger to Microsoft.VSCode.CPP.Extension.linux and dumping a stack trace help you?

**Comment by sean-mcmanus** [comment]
@xueyunquan Which Linux kernel are you using? We've tested with LLVM. I'll check the Chinese pathname later today -- we've definitely tested and fixed bugs in that scenario, but maybe we missed something (https://github.com/Microsoft/vscode-cpptools/issues/870 ).

@hagaishapira Can you provide a stack trace? You can send it to c_cpp_support@microsoft.com if you want. We definitely had users complain about dangling processes a few months ago, we thought we fixed the issues and haven't heard much complaints about that for a while.

**Comment by sean-mcmanus** [comment]
@yyl123hf I'm able to repro the bug with Chinese characters in the path. There was regression in 0.12.2 due to a changed to our implementation of stat on Windows, which caused all the files with non-ASCII characters to be skipped. Thanks a lot for bringing this to our attention. We might track this regression with https://github.com/Microsoft/vscode-cpptools/issues/958, because this bug might have other causes.

**Comment by anhkgg** [comment]
@sean-mcmanus Good job!

**Comment by xueyunquan** [comment]
@sean-mcmanus : This one will do: https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/releases/tag/roc-1.6.1

**Comment by hagaishapira** [comment]
@sean-mcmanus I've sent you guys a stack trace on the email addess you provided. Should you require more sample let me know.

**Comment by hagaishapira** [comment]
Hi,
So this is still recurring and is quite frequent for me even with the latest code and c/c++ 0.12.4.
The fact that Microsoft.VSCode.CPP.IntelliSense.Msvc.linux processes keep hanging and code navigation becomes unresponsive after an hour or so of use is extremely annoying and makes working with code feel very buggy and not mature. 
@sean-mcmanus I've also sent you guys a stack trace by mail for which I have gotten no replies. Would love to provide any sort of assistance to help you guys figure out this bug.

**Comment by sean-mcmanus** [comment]
@hagaishapira The stack trace wasn't helpful. Our extension is still in Preview. There are known bugs which can cause the IntelliSense parser to go into an infinite loop or crash, which could lead to dangling zombie processes. Do you have a repro project available? Does the problem go away if you set the intelliSenseEngine to "Tag Parser"? Maybe we could add code that checks for dead processes and kills them.

**Comment by dcervenkov** [comment]
**Moved to issue #1195**

**Comment by sheinzle** [comment]
I am running into the same issue on Ubuntu 17.04.

After VS Code runs for about 5 minutes, multiple `Microsoft.VSCode.CPP.IntelliSense.Msvc.linux` processes are running and IntelliSense stops working. These processes also survive after closing VS Code, killing the manually from the console is the only thing that helps.

I am happy to provide you with more information if it helps.

- VSCode Version: Code 1.17.2 (b813d12980308015bcd2b3a2f6efa5c810c33ba5, 2017-10-16T13:57:00.652Z)
- OS Version: Linux x64 4.10.0-37-generic
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
python|don|0.7.0
cpptools|ms-|0.14.0
cython|tcw|0.1.0
vscode-todo-highlight|way|0.5.11

**Comment by dcervenkov** [comment]
This has been fixed for me in one of the new versions (not sure which one though). Anyway my previous example works with:
- VSCode Version: Code 1.17.2 (b813d12980308015bcd2b3a2f6efa5c810c33ba5, 2017-10-16T13:57:00.652Z)
- OS Version: Linux x64 4.13.0-16-generic
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
bracket-pair-colorizer|Coe|0.10.10
latex-workshop|Jam|3.6.0
alignment|ann|0.3.0
python|don|0.7.0
cpptools|ms-|0.14.0
vim|vsc|0.10.2

(1 theme extensions excluded)


---


Curiously, this seems to be the same version that gives @sheinzle trouble.

**Comment by sheinzle** [comment]
Thanks @dcervenkov - the only difference is the kernel version. I upgraded my Ubuntu to a non-standard kernel `4.13.10-041310-generic`. For now IntelliSense seems to work, only one `Microsoft.VSCode.CPP.IntelliSense.Msvc.linux` process is running. I will keep you posted.

(BTW, I love VS code so much, that I went through all the trouble of upgrading to a non-standard kernel and finding firmware to get my WIFI working again - big thanks to the whole MS dev team in Zurich)

**Comment by bobbrow** [comment]
@dcervenkov, I've moved your comment into a new issue.  This issue is about Go To Definition.

**Comment by bobbrow** [comment]
@dcervenkov, oh wait, is your issue resolved already?

**Comment by dcervenkov** [comment]
@bobbrow, oh, I should have been more thorough - IIRC both IntelliSense and Go To Definition didn't work before. However, it seems that the problem was solved by a new kernel as observed by @sheinzle (I also moved to a new kernel). It probably still merits looking into as not all of your users will be ready to upgrade their kernels...

**Comment by sheinzle** [comment]
I can confirm that the new kernel removed the problem. IntelliSense & Go To are still working after a full day of coding.

**Comment by krausedj** [comment]
Hey, I still think im seeing this issue, or something similar

Windows 10
VS Code 1.22.2
C/C++ 0.16.1
VIM / Python (Same Behavior when Disabled)
Default Settings

The following code snippet shows some interesting behavior with Intellisense set to the default behavior.
The behavior exists on the variable ThisDoesNotWork, it is unable to locate / peek the declaration or definition.  To me this is interesting, because autocomplete works, and hovering shows the correct definition, but it cannot be located.

```c
#define VAR(vartype, memclass) vartype
#define DUMMY_MEM_CLASS

VAR(int, DUMMY_MEM_CLASS) ThisDoesNotWork;
int ThisDoesWork;

void SomeFunction(void)
{
    ThisDoesNotWork = 1;
    ThisDoesWork = 2;
}
```

Is this a result of the cpp.hint not having my macros in it?
https://github.com/Microsoft/vscode-cpptools/blob/master/Extension/cpp.hint

I tried adding `#define VAR(vartype, memclass) vartype` to the cpp.hint file in my user vscode directory, but still saw the same behavior, perhaps I did it wrong.

**Comment by sean-mcmanus** [comment]
@krausedj Yeah, adding that code to your cpp.hint fixes the issue for me -- the cpp.hint file needs to go in your workspace root, not in the .vscode subfolder. After that, you need to edit/save the file or do a Reset IntelliSense Database.

**Comment by krausedj** [comment]
@sean-mcmanus Thanks, putting it in the workspace root did the trick.

**Comment by sean-mcmanus** [comment]
The dangling processes issues should be fixed as of 0.17.7 (i.e. the cause of the database failures), but you need to make sure no dangling processes from 0.17.6 exist still because we don't terminate the old ones.

**Comment by Traeyee** [comment]
> @hagaishapira You should kill the other extension processes -- the symbol database won't function correctly if multiple processes are trying to access it (i.e. Go to Def would fail). Also, you can try adding a new databaseFilename, such as "${workspaceRoot}/.vscode/vc.db" to have the symbol engine regenerate. Then you can easily see how the size increases.

Just create that file without configuring anything else, right? I don't know is there any need to configure something that points to the 'databaseFilename'.

**Comment by sean-mcmanus** [comment]
@Traeyee Create what file? Ideally, you should kill the dangling processes that are preventing the database from being used (until we figure out a way to deal with this ourselves), but setting/changing the browse.databaseFilename is another fix...our extension creates the database file on our own.
